src/kr_comp.cpp

```diff
@@ -0,0 +1,104 @@
+#include <Rcpp.h>
+#define INCLUDE_RCPP
+#include "utils.h"
+#include "covariance.h"
+#include "chol_derivative.h"
+
+using namespace Rcpp;
+using std::string;
+
+vector<double> as_vector(NumericVector input) {
+  vector<double> ret(as<std::vector<double>>(input));
+  return ret;
+}
+vector<int> as_vector(IntegerVector input) {
+  vector<int> ret(as<std::vector<int>>(input));
+  return ret;
+}
+NumericVector as_nv(vector<double> input) {
+  NumericVector ret(input.size());
+  for (int i = 0; i < input.size(); i++) {
+    ret[i] = input(i);
+  }
+  return ret;
+}
+
+NumericMatrix as_mv(matrix<double> input) {
+  vector<double> input_v = input.vec();
+  NumericVector input_nv = as_nv(input_v);
+  NumericMatrix ret(int(input.rows()), int(input.cols()), input_nv.begin());
+  return ret;
+}
+matrix<double> as_matrix(NumericMatrix input) {
+  matrix<double> ret(input.rows(), input.cols());
+  for (int i = 0; i < input.rows(); i++) {
+    for (int j = 0; j < input.cols(); j++) {
+      ret(i,j) = input(i,j);
+    }
+  }
+  return ret;
+}
+
+List get_pqr(List mmrm_fit, NumericVector theta) {
+  NumericMatrix x = mmrm_fit["x_matrix"];
+  IntegerVector subject_zero_inds = mmrm_fit["subject_zero_inds"];
+  IntegerVector visits_zero_inds = mmrm_fit["visits_zero_inds"];
+  int n_subjects = mmrm_fit["n_subjects"];
+  IntegerVector subject_n_visits = mmrm_fit["subject_n_visits"];
+  int n_visits = mmrm_fit["n_visits"];
+  String cov_type = mmrm_fit["cov_type"];
+  int is_spatial_int = mmrm_fit["is_spatial_int"];
+  bool is_spatial = is_spatial_int == 1;
+  int n_groups = mmrm_fit["n_groups"];
+  IntegerVector subject_groups = mmrm_fit["subject_groups"];
+  NumericVector weights_vector = mmrm_fit["weights_vector"];
+  if (is_spatial) {
+    Rf_error("Spatial Kenward-Roger not implemented yet!");
+  }
+  auto theta_v = as_vector(theta);
+  std::string covtype_str = string(cov_type);
+  auto G_sqrt = as_vector(sqrt(weights_vector));
```

Comment:
the square root issue is here. at first I use:
```{cpp}
auto G = as_vector(weights_vector);
auto Gsqrt = sqrt(G);
// auto Gsqrt = G.sqrt();
// auto Gsqrt = G.cwiseSqrt();
```

it gives seems uninitialized value on my computer, instead of the square root.

Reply:
hm interesting. maybe worth trying to isolate this in a reproducible toy example

Reply:
I tested a bit on total isolation and it seems working fine 
```{cpp}
List my_sqrt(NumericVector x) {
  auto x1 = as_vector(x);
  auto x2 = x1.sqrt();
  auto x3 = x1.cwiseSqrt();
  auto x4 = sqrt(x1);
  return List::create(as_nv(x2), as_nv(x3), as_nv(x4));
}
```

need more investigation here

Reply:
maybe the `as_vector` needs to copy / allocate memory or something like that?

Reply:
@clarkliming ok I guess we shelve this for now...